### PR TITLE
A coalescing operator ?? for Nim

### PR DESCRIPTION
**Pull request by piedar**
The [`coalesce` module](https://github.com/piedar/coalesce) implements a [null coalescing operator](https://en.wikipedia.org/wiki/Null_coalescing_operator) `??` for [Nim](https://github.com/nim-lang/Nim).

## examples

The most obvious case is choosing the first non-nil value.

    let a: string = nil
    let x = a ?? "b"
    assert x == "b"

The operator also supports `Option[T]`.

    let a: Option[string] = none(string)
    let x = a ?? "b"
    assert x == "b"

Longer chains work too, and the expression short-circuits if possible.

    let result = tryComputeFast(input) ?? tryComputeSlow(input) ?? myDefault

## todo

- support for `not nil`

### CODE REVIEW

**Comment by krux02** [comment]
Just a warning, you introduced a new operator. Don't expect this to be pulled simply.

For the Option type, this alreay exists, it is just called get and it is not at operator.
```nim
import options

let a: Option[string] = none(string)
let x = a.get("b")
assert x == "b"
```
A pull request that makes ``option.get`` a template would be less intrusive as a pull request, and it would allow this to work:

``let result = tryComputeFast(input).get(tryComputeSlow(input)).get(myDefault)``


A current goal for string/seq type is that they won't allow the nil value anymore, so I don't think improving the usability of nilable seq/vector values is a good idea.

**Comment by zielmicha** [comment]
Why not make it a Nimble package?

**Comment by krux02** [comment]
too small for a nimble package. Too much of a change for the standard library. There is nothing wrong when someone wants to improve the standard library. But introducing new operators is not the way to solve things. Nim has already too many operators and to my taste Nim should deprecate all operators except the base operators and provide proper names for all operations. There are use cases for new operators, but they are rare.

**Comment by piedar** [comment]
@zielmicha There is a nimble package at [piedar/coalesce](https://github.com/piedar/coalesce), but someone suggested that I propose it for inclusion in the `options` module.

@krux02 I'm glad to hear that strings may become non-nilable!  I'm hoping there will also be improvements to the `not nil` checker, for I have found it to be too easily fooled.  I will open a separate PR to make `option.get(T otherwise)` a template (despite my opinion that its syntax is confusing).

**Comment by krux02** [comment]
Well I think it will develop similar to Rust and Scala. Nil will not disappear, but you would need to work hard to actually get nil values. Things will not be Nil by default anymore.

**Comment by zielmicha** [comment]
`option.get(otherwise)` should not be made a template. In would be very confusing to have it not evaluate `otherwise` in some cases. 

`??` is already used in C#, so I think its intent will be pretty clear without specialized name. Another option would be to use `or` operator (as in Python).

Personally I would vote for naming it `or` and including it in `system` module (for `ref` and `ptr`) and `options` (for `Option`).

**Comment by piedar** [comment]
Does Nim have something like Python's [Truth Value Testing](https://docs.python.org/3/library/stdtypes.html#truth-value-testing)?  If so, `nil` and `none(T)` could be considered false and then `or` would work great as a coalescing operator.

**Comment by TiberiumN** [comment]
@piedar no, because it's too implicit.

**Comment by dom96** [comment]
So this is pretty cool. I'm unsure whether it should make it into the stdlib, this is practically a language-wide change so I want to see what @Araq thinks.

I will review your PR in any case.

File: tests/options/tcoalesce.nim [hunk 1]
```diff
@@ -0,0 +1,154 @@
+import options, unittest
+
+suite "coalesce":
```
Comment (dom96) [inline tests/options/tcoalesce.nim]: The tests should be in a single ``toptions`` module inside ``tests/stdlib`` (or at the bottom of the ``options`` module).

File: lib/pure/options.nim [hunk 2]
```diff
@@ -153,70 +153,24 @@ proc `$`*[T]( self: Option[T] ): string =
...
-    test "map":
-      check( some(123).map(proc (v: int): int = v * 2) == some(246) )
-      check( intNone.map(proc (v: int): int = v * 2).isNone )
-
-    test "filter":
-      check( some(123).filter(proc (v: int): bool = v == 123) == some(123) )
-      check( some(456).filter(proc (v: int): bool = v == 123).isNone )
-      check( intNone.filter(proc (v: int): bool = check false).isNone )
+template `??`*[T](left: T, right: T): T =
+  ## Returns ``left`` if not nil or ``right`` otherwise.
```
Comment (dom96) [inline lib/pure/options.nim]: Some examples in the documentation would be helpful, search for ``code-block:`` to see how to embed source code into the docs.

**Comment by Araq** [comment]
We plan to make ``nil`` a rare thing so I vote against it. Rare things do not need syntactic shortcuts.

**Comment by piedar** [comment]
Thanks for the comments @dom96!  I'll keep those in mind for when we move forward.  I'm totally fine with this staying as a nimble package in the meantime.

Since @Araq suggests to avoid (ab)use of `nil`, let's consider only the two templates with `Option[T]` on the left side.  I think the operator can still be valuable in that more limited scope.  Thoughts?

```
# word vs symbol
tree = makeFir() or makeOak() or makePalm();
tree = makeFir() ?? makeOak() ?? makePalm();
```

**Comment by vegansk** [comment]
Maybe we must implement something like haskell's do notation in the standard library ([example](https://github.com/vegansk/nimfp/blob/master/tests/fp/test_forcomp.nim#L48-L59)) that works with many types, instead of implementing different syntaxes for the same things but for different types: coalescing operator for option, await for futures etc.

For example, do notation from nimfp library works with Option, List, Either, etc. It works even with Future from stdlib.

**Comment by Araq** [comment]
> Since @Araq suggests to avoid (ab)use of nil, let's consider only the two templates with Option[T] on the left side. I think the operator can still be valuable in that more limited scope. Thoughts?

I think I like the idea.

**Comment by Araq** [comment]
No feedback. Closing.

**Pull request closed without merging**